plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Completion/Completer/FilePath.hs

```diff
@@ -0,0 +1,190 @@
+              pure $ if isDir then Posix.addTrailingPathSeparator d else d
+          )
+          dirs
+      pure fixedDirs
+    Left (err :: IOError) -> do
+      logWith recorder Warning $ LogFilePathCompleterIOError complDir err
+      pure []
+
+{- | Returns a list of all (and only) directories in the
+  directory described by path completion info.
+-}
+listDirectoryCompletions :: Recorder (WithPriority Log) -> PathCompletionInfo -> IO [FilePath]
+listDirectoryCompletions recorder complInfo = do
+  filepaths <- listFileCompletions recorder complInfo
+  filterM (doesDirectoryExist . mkDirFromCWD complInfo) filepaths
+
+pathCompletionInfoFromCabalPrefixInfo :: CabalPrefixInfo -> PathCompletionInfo
+pathCompletionInfoFromCabalPrefixInfo ctx =
+  PathCompletionInfo
+    { partialFileName = dirNamePrefix
+    , partialFileDir = Posix.addTrailingPathSeparator $ Posix.takeDirectory prefix
+    , workingDir = dir
+    }
+ where
+  prefix = T.unpack $ completionPrefix ctx
+  dirNamePrefix = T.pack $ Posix.takeFileName prefix
+  dir = completionWorkingDir ctx
+
+{- | Returns the directory, the currently handled cabal file is in.
+
+  We let System.FilePath handle the separator syntax since this is used
+  to query filepaths from the system. See Note [Using correct file path separators].
+-}
+mkCompletionDirectory :: PathCompletionInfo -> FilePath
+mkCompletionDirectory complInfo =
+  FP.addTrailingPathSeparator $
+    workingDir complInfo FP.</> (FP.normalise $ partialFileDir complInfo)
+
+{- | Returns the complete filepath for the given partial filepath
+  by combining the current working directory of the cabal file
+  with the given partly written file path.
+
+  Since this is used for completions we use posix separators here.
+  See Note [Using correct file path separators].
+-}
+mkDirFromCWD :: PathCompletionInfo -> FilePath -> FilePath
+mkDirFromCWD complInfo fp =
+  Posix.addTrailingPathSeparator $
+    mkCompletionDirectory complInfo Posix.</> Posix.normalise fp
+
+{- | Takes a directory and PathCompletionInfo and
+  returns the whole path including the prefix that was already written.
+
+  Since this is used for completions we use posix separators here.
+  See Note [Using correct file path separators].
+-}
+mkPathCompletion :: PathCompletionInfo -> T.Text -> T.Text
+mkPathCompletion complInfo completion =
+  T.pack $
+    partialFileDir complInfo Posix.</> T.unpack completion
+
+{-   Takes a suffix, a completed path and a pathCompletionInfo and
+   generates the whole filepath including the already written prefix,
+   and the suffix in case the completed path is a filepath.
+-}
+mkFilePathCompletion :: T.Text -> T.Text -> PathCompletionInfo -> IO T.Text
+mkFilePathCompletion suffix completion complInfo = do
+  let combinedPath = T.unpack $ mkPathCompletion complInfo completion
+  isFilePath <- doesFileExist combinedPath
+  let completedPath = if isFilePath then combinedPath ++ T.unpack suffix else combinedPath
```

Comment:
... why? I see what this is doing but I can't really figure out why this is what you want :thinking: